### PR TITLE
Fix #114: Add Knapsack model

### LINKED ISSUE #114: [Model] Knapsack
## Motivation

One of Karp's 21 NP-complete problems; foundational in combinatorial optimization, cryptography, and resource allocation. Has direct reductions to ILP and QUBO.

## Definition

**Name:** Knapsack
**Reference:** [Karp, 1972](https://doi.org/10.1007/978-1-4684-2001-2_9); [Kellerer, Pferschy & Pisinger, *Knapsack Problems*, 2004](https://doi.org/10.1007/978-3-540-24777-7)

Given $n$ items, each with weight $w_i$ and value $v_i$, and a capacity $C$, find a subset $S \subseteq \{0, \ldots, n-1\}$ such that $\sum_{i \in S} w_i \leq C$, maximizing $\sum_{i \in S} v_i$.

## Variables

- **Count:** $n$ (one variable per item)
- **Per-variable domain:** binary $\{0, 1\}$
- **Meaning:** $x_i = 1$ if item $i \in S$ (selected)

## Schema (data type)

**Type name:** Knapsack
**Variants:** 0-1 (binary), bounded, unbounded

| Field | Type | Description |
|-------|------|-------------|
| weights | `Vec<i64>` | Item weights $w_0, \ldots, w_{n-1}$ |
| values | `Vec<i64>` | Item values $v_0, \ldots, v_{n-1}$ |
| capacity | `i64` | Knapsack capacity $C$ |

## Problem Size

| Metric | Expression | Description |
|--------|------------|-------------|
| num_items | $n$ | Number of items |

## Complexity

- **Decision complexity:** NP-complete (Karp, 1972)
- **Best known exact algorithm:** $O(nC)$ pseudo-polynomial dynamic programming; $O(2^{n/2})$ via meet-in-the-middle
- **Approximation:** FPTAS exists — $(1-\varepsilon)$-approximation in $O(n^2 / \varepsilon)$ time
- **References:** [Karp 1972](https://doi.org/10.1007/978-1-4684-2001-2_9); [Ibarra & Kim 1975 (FPTAS)](https://doi.org/10.1145/321906.321909)

## Extra Remark

Knapsack is weakly NP-hard (admits pseudo-polynomial algorithms), unlike strongly NP-hard problems such as TSP. It is a special case of ILP with a single constraint. The subset-sum problem is the special case where $v_i = w_i$. Knapsack-based cryptosystems (Merkle-Hellman) were historically important in public-key cryptography.

## How to solve

- [x] It can be solved by (existing) bruteforce.
- [ ] It can be solved by reducing to integer programming, through a new Knapsack → ILP rule issue (to be filed).

Bruteforce: enumerate all $2^n$ subsets, check capacity constraint, return maximum value.

## Example Instance

$n = 4$ items, capacity $C = 7$:

| Item | Weight | Value | Value/Weight |
|:---:|:---:|:---:|:---:|
| 0 | 2 | 3 | 1.50 |
| 1 | 3 | 4 | 1.33 |
| 2 | 4 | 5 | 1.25 |
| 3 | 5 | 7 | 1.40 |

Feasible subsets:

| Subset | Weight | Value |
|:---:|:---:|:---:|
| $\{0, 3\}$ | 7 | **10** |
| $\{1, 2\}$ | 7 | 9 |
| $\{0, 2\}$ | 6 | 8 |
| $\{0, 1\}$ | 5 | 7 |
| $\{3\}$ | 5 | 7 |

Optimal: $S = \{0, 3\}$, $x = (1,0,0,1)$, weight $= 7$, value $= 10$.

Note: greedy by value/weight ratio picks item 0 (1.50) then item 3 (1.40), which happens to be optimal here, but greedy is not optimal in general.

### PR DESCRIPTION
**Pull request by zazabap**
## Summary

- Add the 0-1 Knapsack problem model (`src/models/misc/knapsack.rs`)
- Binary configuration space: select/don't select each item
- Maximize total value subject to weight capacity constraint
- Complexity: O(2^(n/2)) via Horowitz-Sahni meet-in-the-middle (1974)
- 15 unit tests covering evaluation, edge cases (zero capacity, single item, greedy-not-optimal adversarial), brute force solver, serialization
- CLI dispatch + alias (`KS`)
- Paper problem-def entry with formal definition and background
- Schema export updated

## Test plan

- [x] All 15 unit tests pass (`cargo test -- knapsack`)
- [x] Doc-test passes
- [x] Clippy clean (no warnings)
- [x] Structural review: 16/16 checks pass
- [x] Issue compliance: 8/8 checks pass
- [x] Code quality review: no critical or important issues

Closes #114

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: problemreductions-cli/src/problem_name.rs [hunk 1]
```diff
@@ -21,6 +21,7 @@ pub const ALIASES: &[(&str, &str)] = &[
     ("TSP", "TravelingSalesman"),
     ("BP", "BinPacking"),
     ("CVP", "ClosestVectorProblem"),
+    ("KS", "Knapsack"),
```
[the remark this answers is not in the document]
Comment (zazabap) [reply problemreductions-cli/src/problem_name.rs]: Removed both `KS` and `BP` aliases (from ALIASES constant and resolve_alias match arms). Case-insensitive full names (`knapsack`, `binpacking`) still work.

File: docs/paper/reductions.typ [hunk 2]
```diff
@@ -886,6 +887,14 @@ Biclique Cover is equivalent to factoring the biadjacency matrix $M$ of the bipa
...
+][
+  One of Karp's 21 NP-complete problems @karp1972. Knapsack is only _weakly_ NP-hard: a classical dynamic-programming algorithm runs in $O(n C)$ pseudo-polynomial time, and a fully polynomial-time approximation scheme (FPTAS) achieves $(1 - epsilon)$-optimal value in $O(n^2 slash epsilon)$ time @ibarra1975. The special case $v_i = w_i$ for all $i$ is the Subset Sum problem. Knapsack is also a special case of Integer Linear Programming with a single constraint. The best known exact algorithm is the $O^*(2^(n slash 2))$ meet-in-the-middle approach of Horowitz and Sahni @horowitz1974, which partitions items into two halves and combines sorted sublists.
+
+  *Example.* Let $n = 4$ items with weights $(2, 3, 4, 5)$, values $(3, 4, 5, 7)$, and capacity $C = 7$. Selecting $S = {1, 2}$ (items with weights 3 and 4) gives total weight $3 + 4 = 7 lt.eq C$ and total value $4 + 5 = 9$. Selecting $S = {0, 3}$ (weights 2 and 5) gives weight $2 + 5 = 7 lt.eq C$ and value $3 + 7 = 10$, which is optimal.
```
[the remark this answers is not in the document]
Comment (zazabap) [reply docs/paper/reductions.typ]: Fixed: changed to consistent 0-based indexing (`w_0, ..., w_(n-1)`) to match the subset `S ⊆ {0, ..., n-1}` and the code implementation (configs are 0-indexed).

**Review by zazabap — COMMENTED** [review]
Addressed: the paper definition now uses consistent 0-based indexing (`w_0, ..., w_(n-1)`, `v_0, ..., v_(n-1)`) matching the subset `S ⊆ {0, ..., n-1}` and the code's 0-indexed configs. Fixed in commit 48239c5.

**Review by GiggleLiu — APPROVED** [review]
